cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -40,36 +65,109 @@ class device_memory_resource : public resource {
+  }
 };
 
 /**
  * Load a temp workspace resource from a resources instance (and populate it on the res
  * if needed).
+ *
  * @param res raft resources object for managing resources
  * @return device memory resource object
  */
-inline rmm::mr::device_memory_resource* get_workspace_resource(resources const& res)
+inline auto get_workspace_resource(resources const& res)
+  -> rmm::mr::limiting_resource_adaptor<rmm::mr::device_memory_resource>*
 {
   if (!res.has_resource_factory(resource_type::WORKSPACE_RESOURCE)) {
     res.add_resource_factory(std::make_shared<workspace_resource_factory>());
   }
-  return res.get_resource<rmm::mr::device_memory_resource>(resource_type::WORKSPACE_RESOURCE);
+  return res.get_resource<rmm::mr::limiting_resource_adaptor<rmm::mr::device_memory_resource>>(
+    resource_type::WORKSPACE_RESOURCE);
+};
+
+/**
+ * Set a temporary workspace resource on a resources instance.
+ *
+ * @param res raft resources object for managing resources
+ * @param mr an optional RMM device_memory_resource
+ * @param allocation_limit
+ *   the total amount of memory in bytes available to the temporary workspace resources.
+ * @param alignment optional alignment requirements passed to RMM allocations
+ *
+ */
+inline void set_workspace_resource(resources const& res,
+                                   std::shared_ptr<rmm::mr::device_memory_resource> mr = {nullptr},
+                                   std::optional<std::size_t> allocation_limit = std::nullopt,
+                                   std::optional<std::size_t> alignment        = std::nullopt)
+{
+  res.add_resource_factory(
+    std::make_shared<workspace_resource_factory>(mr, allocation_limit, alignment));
 };
 
 /**
- * Set a temp workspace resource on a resources instance.
+ * Set a temporary workspace resource on a resources instance.
  *
  * @param res raft resources object for managing resources
- * @param mr a valid rmm device_memory_resource
+ * @param mr an optional RMM device_memory_resource;
+ *   note, the ownership of the object is not transferred with this raw pointer interface.
+ * @param allocation_limit
+ *   the total amount of memory in bytes available to the temporary workspace resources.
+ * @param alignment optional alignment requirements passed to RMM allocations
+ *
  */
-inline void set_workspace_resource(resources const& res, rmm::mr::device_memory_resource* mr)
+inline void set_workspace_resource(resources const& res,
+                                   rmm::mr::device_memory_resource* mr,
+                                   std::optional<std::size_t> allocation_limit = std::nullopt,
+                                   std::optional<std::size_t> alignment        = std::nullopt)
 {
-  res.add_resource_factory(std::make_shared<workspace_resource_factory>(mr));
+  // NB: to preserve the semantics of passing memory resource without transferring the ownership,
+  //     we create a shared pointer with a dummy deleter (void_op).
```

Comment:
Is there any reason to wrap it into a `shared_ptr `in this case?

Reply:
Yes, I opted for the `shared_ptr` semantics in `limited_memory_resource` to make it easier for the end-user to pass custom memory resources (e.g. create a pool memory resource and not worry about managing its lifetime). Since the  `limited_memory_resource` accepts only shared pointers, we need to wrap the raw pointer.

There's only one exception to this whole story - `rmm::mr::get_current_device_resource()`, which returns a pointer to the memory resource that is manages elsewhere. The original workspace resource API was tailored for this use case; to keep the PR non-breaking, I decided to leave all signatures as they were, hence the shared_ptr with a dummy deleter.

Reply:
I know we were accepting raw pointers in some places already but shared_ptr does provide a great mechanism for unified already owned elsewhere pointers vs one being owned by the shared pointer itself. What do you say we just accept shared_ptr everywhere for the memory resources to make the api experience more consistent?

Reply:
I'm keeping the raw-pointer version here for two reasons:
 1. Just to keep the PR 'non-breaking'
 2. To allow a user to pass `rmm::mr::get_current_device_resource()` that is managed statically by rmm (though this can be workarounded by wrapping it into a shared_ptr with a dummy deleter).

@cjnolet let me know if you'd like to remove this, I don't mind.

Reply:
I think we are so early in the process of using the workspace resource that it would be a good time to clean this up and make it consistent. I doubt anyone is using it downstream, really, so I'm okay changing it here.